### PR TITLE
chore: move app URL parsing to its own package

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

Parsing `CODER_WILDCARD_ACCESS_URL` as a `url.URL` was always incorrect (see https://goplay.tools/snippet/YHUpcSkVrli). Since wildcard urls do not have a scheme, the url parsing was never accurate to how it parsed things. And it would not accept ports.

We have a wildcard parse function, so why not make that the validation instead of using the url and `url.String()`.

Lastly I moved `appurl` to it's own package to avoid import loops. This is a reasonable thing to package.

In support of https://github.com/coder/coder/issues/8189. We need to move the field from a url to a string to support ports.

### CODE REVIEW

**Comment by Emyrk** [comment]
* `main`
  * **#11651** <a href="https://app.graphite.dev/github/pr/coder/coder/11651?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Emyrk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: coderd/workspaceapps/appurl/appurl.go [hunk 1]
```diff
@@ -10,8 +10,8 @@ import (
 )
 
 var (
-	// Remove the "starts with" and "ends with" regex components.
-	nameRegex = strings.Trim(UsernameValidRegex.String(), "^$")
+	// nameRegex is the same as our UsernameRegex without the ^ and $.
+	nameRegex = "[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)*"
```
Comment (deansheather) [inline coderd/workspaceapps/appurl/appurl.go:14]: I think it should still trim the old regex instead TBH

**Review by deansheather — APPROVED** [review]

File: coderd/workspaceapps/appurl/appurl.go (hunk 1 above)
Comment (Emyrk) [reply coderd/workspaceapps/appurl/appurl.go:14]: Because of import loops, we would need to put that regex in some other package then :cry:.

Comment (deansheather) [reply coderd/workspaceapps/appurl/appurl.go:14]: codersdk maybe?

**Pull request merged**